### ISSUE TITLE
Importing subscribers who were initially blocklisted during a previous import

### ISSUE THREAD

**Issue by Kannaj** [knadh/listmonk#929]
Hi Kailash,

**TLDR**: we see a bug when we upload a bunch of users (previously blocklisted) to a new list. Ideally, they should still be updated as 'unsubscribed' in the new list 

-------

We recently moved our newsletter arch from email octopus to listmonk. This process involved generating a 'subscribed' list and an 'unsubscribe' list from email octopus. FYI, these lists are not mutually exclusive and there is some overlap.

Our import procedure involved the below

1. Import the new list of subscribers as `confirmed` in `/admin/subscribers/import`
2. Blocklist a portion of those users (and add additional ones)  

The problem happens when we import a new list of subscribers from our prod db and if those users happen to already be marked as blocklisted in listmonk db.

These users are marked as 'confirmed' in the `subscriber_list` table but are `blocklisted` at the subscriber level. This means that they're essentially counted in the `subIDs` CTE but filtered out in the `subs` CTE

```
WITH camps AS (
    SELECT last_subscriber_id, max_subscriber_id, type FROM campaigns WHERE id = $1 AND status='running'
),
campLists AS (
    SELECT lists.id AS list_id, optin FROM lists
    LEFT JOIN campaign_lists ON (campaign_lists.list_id = lists.id)
    WHERE campaign_lists.campaign_id = $1
),
subIDs AS (
    SELECT DISTINCT ON (subscriber_lists.subscriber_id) subscriber_id, list_id, status FROM subscriber_lists
    WHERE
        -- ARRAY_AGG is 20x faster instead of a simple SELECT because the query planner
        -- understands the CTE's cardinality after the scalar array conversion. Huh.
        list_id = ANY((SELECT ARRAY_AGG(list_id) FROM campLists)::INT[]) AND
        status != 'unsubscribed' AND
        subscriber_id > (SELECT last_subscriber_id FROM camps) AND
        subscriber_id <= (SELECT max_subscriber_id FROM camps)
    ORDER BY subscriber_id LIMIT $2
),
subs AS (
    SELECT subscribers.* FROM subIDs
    LEFT JOIN campLists ON (campLists.list_id = subIDs.list_id)
    INNER JOIN subscribers ON (
        subscribers.status != 'blocklisted' AND
        subscribers.id = subIDs.subscriber_id AND

        (CASE
            -- For optin campaigns, only e-mail 'unconfirmed' subscribers.
            WHEN (SELECT type FROM camps) = 'optin' THEN subIDs.status = 'unconfirmed' AND campLists.optin = 'double'

            -- For regular campaigns with double optin lists, only e-mail 'confirmed' subscribers.
            WHEN campLists.optin = 'double' THEN subIDs.status = 'confirmed'

            -- For regular campaigns with non-double optin lists, e-mail everyone
            -- except unsubscribed subscribers.
            ELSE subIDs.status != 'unsubscribed'
        END)
    )
),
u AS (
    UPDATE campaigns
    SET last_subscriber_id = (SELECT MAX(id) FROM subs),
        sent = sent + (SELECT COUNT(id) FROM subs),
        updated_at = NOW()
    WHERE (SELECT COUNT(id) FROM subs) > 0 AND id=$1
)
SELECT * FROM subs;
```

This is sometimes causing some weird issues where a campaign only sends emails to a subsection of the users in any given list. (if you look at the first 3 campaigns - 0/3k emails were sent)

![image](https://user-images.githubusercontent.com/13432536/187425927-e1482284-9a06-49e4-a7f8-04c1034e1f26.png)

To fix this, I manually run a query in the listmonk DB with something like the below

```
with active_blocklisted_subs as (
	select s.id
	from subscribers s 
	left join subscriber_lists sl on s.id = sl.subscriber_id 
	where sl.list_id is not null and sl.status  != 'unsubscribed' and s.status = 'blocklisted'
)
update subscriber_lists set status = 'unsubscribed', updated_at=NOW() where subscriber_id IN (select * from active_blocklisted_subs)
```

This ensures that all blocklisted users are updated as 'unsubscribed' in all the lists they belong to

im attaching a few sample csvs which could help you replicate this issue

[list_1_blocklist.csv](https://github.com/knadh/listmonk/files/9452395/list_1_blocklist.csv)
[list_1_subscribe.csv](https://github.com/knadh/listmonk/files/9452396/list_1_subscribe.csv)
[list_2_subscribe.csv](https://github.com/knadh/listmonk/files/9452397/list_2_subscribe.csv)

1. Upload `list_1_subscribe.csv` to a default list etc (userids 1-5000)
2. Upload  `list_1_blocklist.csv` as blocklisted users (blocklists the first 1500 users)
3. Upload `list_2_subscribe.csv` to a new list (say list_2) - reuploads the user list with ids from (0-1000)

If you can send a new campaign to the users of `list_2` - the campaigns just stops without sending any emails.

is there any official way to fix this? I am guessing we can tweak this SQL query somehow to always add a user as `unsubscribed` if he was previously blocklisted. 

```
-- name: upsert-subscriber
-- Upserts a subscriber where existing subscribers get their names and attributes overwritten.
-- If $7 = true, update values, otherwise, skip.
WITH sub AS (
    INSERT INTO subscribers as s (uuid, email, name, attribs, status)
    VALUES($1, $2, $3, $4, 'enabled')
    ON CONFLICT (email)
    DO UPDATE SET
        name=(CASE WHEN $7 THEN $3 ELSE s.name END),
        attribs=(CASE WHEN $7 THEN $4 ELSE s.attribs END),
        updated_at=NOW()
    RETURNING uuid, id
),
subs AS (
    INSERT INTO subscriber_lists (subscriber_id, list_id, status)
    VALUES((SELECT id FROM sub), UNNEST($5::INT[]), $6)
    ON CONFLICT (subscriber_id, list_id) DO UPDATE
    SET updated_at=NOW(), status=(CASE WHEN $7 THEN $6 ELSE subscriber_lists.status END)
)
SELECT uuid, id from sub;
```

**Comment by knadh** [comment]
This is a bit of a tricky one. Users are blocklisted when they exercise the `do not e-mail me ever again` option. An import or a sync overriding that by default is not desirable. If anything, this should be a new explicit option on the import page.

**Comment by Kannaj** [comment]
if I'm understanding this correctly - are you talking about adding an option/checkbox like 'ignore blocklisted users' to the import page?

**Comment by knadh** [comment]
Blocklisted subscribers are already ignored. You're referring to re-activating them after an import, right? Such an option would be a checkbox that says ~ `Override blocklisted users too`.
